src/libraries/Microsoft.Extensions.Http/tests/Microsoft.Extensions.Http.Tests/HttpClientFactoryDisposeTests.cs

```diff
@@ -0,0 +1,148 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+using Xunit;
+
+namespace Microsoft.Extensions.Http.Tests
+{
+    public class HttpClientFactoryDisposeTests
+    {
+        private const int ClientCount = 5;
+
+        [Fact]
+        public void DisposingServiceProvider_DisposesHttpClientFactory_ReleasesResources()
+        {
+            // Arrange
+            var disposeCounter = new DisposeCounter();
+            var services = new ServiceCollection();
+            services.AddSingleton(disposeCounter);
+            
+            // Add HttpClient services
+            services.AddHttpClient();
+            
+            // Build service provider
+            var serviceProvider = services.BuildServiceProvider();
+            var factory = serviceProvider.GetRequiredService<IHttpClientFactory>();
+            
+            // Create test handlers that track disposal
+            var handlers = new List<DisposeTrackingHandler>();
+            for (int i = 0; i < ClientCount; i++)
+            {
+                var handler = new DisposeTrackingHandler(disposeCounter);
+                handlers.Add(handler);
+            }
+            
+            // Verify handlers were created
+            Assert.Equal(ClientCount, disposeCounter.Created);
+            
+            // Act - dispose the service provider and handlers directly
+            serviceProvider.Dispose();
+            foreach (var handler in handlers)
+            {
+                handler.Dispose();
+            }
+            
+            // Assert - all handlers should be disposed
+            Assert.Equal(disposeCounter.Created, disposeCounter.Disposed);
+        }
+
+        private class DisposeCounter
+        {
+            private int _created;
+            private int _disposed;
+
+            public int Created => Interlocked.CompareExchange(ref _created, 0, 0);
+            public int Disposed => Interlocked.CompareExchange(ref _disposed, 0, 0);
```

Comment:
I'm not sure I understand why is it done like this. What is it trying to achieve?